tpm2/src/util/tpm2_kmyth_seal.c

```diff
@@ -30,12 +30,443 @@ extern const cipher_t cipher_list[];
+
+    // clear potential 'auth' data, free TPM resources before exiting early
+    kmyth_clear(objAuthVal.buffer, objAuthVal.size);
+    kmyth_clear(ownerAuth.buffer, ownerAuth.size);
+    tpm2_free_resources(&sapi_ctx);
+    return 1;
+  }
+  kmyth_log(LOG_DEBUG, "retrieved SRK handle (0x%08X)", storageRootKey_handle);
+
+  // We create a storage key (SK) that we will use to seal a symmetric
+  // wrapping key that we will create and use to encrypt the user input data.
+  // This storage key will be sealed to the SRK (its parent is the SRK).
+  TPM2B_PRIVATE storageKey_private;
+  TPM2B_PUBLIC storageKey_public;
+  TPM2_HANDLE storageKey_handle = 0;
+
+  storageKey_private.size = 0;
+  storageKey_public.size = 0;
+  if (tpm2_kmyth_create_sk(sapi_ctx,
+                           storageRootKey_handle,
+                           ownerAuth,
+                           objAuthVal,
+                           objPcrList, objAuthPolicy,
+                           &storageKey_handle,
+                           &storageKey_private, &storageKey_public))
+  {
+    kmyth_log(LOG_ERR, "failed to create a storage key ... exiting");
+
+    // clear potential 'auth' data, free TPM resources before exiting early
+    kmyth_clear(objAuthVal.buffer, objAuthVal.size);
+    kmyth_clear(ownerAuth.buffer, ownerAuth.size);
+    tpm2_free_resources(&sapi_ctx);
+    return 1;
+  }
+
+  // As this newly created storage key will be used by the TPM, we must load it
+  SESSION *nullAuthSession = NULL;  // no policy to auth load into SRK hierarchy
+  TPML_PCR_SELECTION emptyPcrList;
+
+  emptyPcrList.count = 0;       // no auth policy session means no PCR criteria
+  if (tpm2_kmyth_load_object(sapi_ctx,
+                             nullAuthSession,
+                             storageRootKey_handle,
+                             ownerAuth,
+                             emptyPcrList,
+                             &storageKey_private,
+                             &storageKey_public, &storageKey_handle))
+  {
+    kmyth_log(LOG_ERR, "failed to load storage key ... exiting");
+
+    // clear potential 'auth' data, free TPM resources before exiting early
+    kmyth_clear(objAuthVal.buffer, objAuthVal.size);
+    kmyth_clear(ownerAuth.buffer, ownerAuth.size);
+    tpm2_free_resources(&sapi_ctx);
+    return 1;
+  }
+
+  // Done with owner hierarchy authorization - SRK and SK available in TPM
+  kmyth_clear(ownerAuth.buffer, ownerAuth.size);
+
+  // Wrap input data -
+  //   - The data to be encrypted is contained in a file and the path to that
+  //     file is specified by the user.
+  //   - The encryption uses the symmetric 'cipher' specified by the user.
+  //   - The symmetric wrapping key used for encryption
+  //     is created as part of the call to kmyth_wrap_input().
+  kmyth_log(LOG_DEBUG, "wrapping input data");
+  size_t wrapped_data_size = 0;
+  unsigned char *wrapped_data = NULL;
+  size_t wrapKey_size = get_key_len_from_cipher(cipher) / 8;
```

Comment:
Do we have any restrictions on cipher key lengths? Here we seem to assume they're some number of full octets, but what happens if somebody tries to use a cipher whose key length in bits isn't a multiple of 8?